sdk/opa.go

```diff
@@ -0,0 +1,426 @@
+	interQueryBuiltinCache cache.InterQueryCache
+	queryCache             *queryCache
+}
+
+// New returns a new OPA object. This function should minimally be called with
+// options that specify an OPA configuration file.
+func New(ctx context.Context, opts Options) (*OPA, error) {
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := opts.init(); err != nil {
+		return nil, err
+	}
+
+	opa := &OPA{
+		id: id,
+		state: &state{
+			queryCache: newQueryCache(),
+		},
+	}
+
+	opa.config = opts.config
+	opa.logger = opts.Logger
+	opa.console = opts.ConsoleLogger
+
+	return opa, opa.configure(ctx, opa.config, opts.Ready, opts.block)
+}
+
+// Configure updates the configuration of the OPA in-place. This function should
+// be called in response to configuration updates in the environment. This
+// function is atomic. If the configuration update cannot be successfully
+// applied, the old configuration will remain intact.
+func (opa *OPA) Configure(ctx context.Context, opts ConfigOptions) error {
+
+	if err := opts.init(); err != nil {
+		return err
+	}
+
+	// NOTE(tsandall): In future we could be more intelligent about
+	// re-configuration and avoid expensive background processing.
+	opa.mtx.Lock()
+	equal := bytes.Equal(opts.config, opa.config)
+	opa.mtx.Unlock()
+
+	if equal {
+		close(opts.Ready)
+		return nil
+	}
+
+	return opa.configure(ctx, opts.config, opts.Ready, opts.block)
+}
+
+func (opa *OPA) configure(ctx context.Context, bs []byte, ready chan struct{}, block bool) error {
+
+	manager, err := plugins.New(
+		bs,
+		opa.id,
+		inmem.New(),
+		plugins.Logger(opa.logger),
+		plugins.ConsoleLogger(opa.console))
+	if err != nil {
+		return err
+	}
+
+	manager.RegisterCompilerTrigger(func(_ storage.Transaction) {
+		opa.mtx.Lock()
+		opa.state.queryCache.Clear()
+		opa.mtx.Unlock()
+	})
+
+	var mtx sync.Mutex
+	var done bool
+
+	manager.RegisterPluginStatusListener("sdk", func(status map[string]*plugins.Status) {
+
+		mtx.Lock()
+		defer mtx.Unlock()
+
+		if done {
+			return
+		}
+
+		// NOTE(tsandall): we do not include a special case for the discovery
+		// plugin. If the discovery plugin is the only plugin and it goes ready,
+		// then OPA will be considered ready. The discovery plugin only goes ready
+		// _after_ it has successfully processed a discovery bundle. During
+		// discovery bundle processing, other plugins will register so their states
+		// will be accounted for. If a discovery bundle did not enable any other
+		// plugins (bundles, etc.) the OPA will still be operational.
+		for _, s := range status {
+			if s.State != plugins.StateOK {
+				return
+			}
+		}
+
+		close(ready)
+		done = true
+	})
+
+	d, err := discovery.New(manager)
+	if err != nil {
+		return err
+	}
+
+	manager.Register(discovery.Name, d)
+
+	if err := manager.Start(ctx); err != nil {
+		return err
+	}
+
+	if block {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ready:
+			break
+		}
```

Comment:
```suggestion
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ready:
		}
```
Works the same, I think?

Reply:
👍